extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcIdentityProvider.java

```diff
@@ -81,101 +87,157 @@ public SecurityIdentity get() {
+        final TokenCredential tokenCred = request.getToken();
+        if (tokenCred instanceof AccessTokenCredential && ((AccessTokenCredential) tokenCred).isOpaque()) {
+            // remote introspection is required, a blocking call
+            if (BlockingOperationControl.isBlockingAllowed()) {
+                return Uni.createFrom().emitter(new Consumer<UniEmitter<? super SecurityIdentity>>() {
+                    @Override
+                    public void accept(UniEmitter<? super SecurityIdentity> uniEmitter) {
+                        createSecurityIdentityWithOidcServer(uniEmitter, vertxContext, request, resolvedContext, userInfo);
+                    }
+                });
+            } else {
+                return Uni.createFrom().emitter(
+                        new Consumer<UniEmitter<? super SecurityIdentity>>() {
                             @Override
-                            public void handle(AsyncResult<AccessToken> event) {
-                                if (event.failed()) {
-                                    uniEmitter.fail(new AuthenticationFailedException(event.cause()));
-                                    return;
-                                }
+                            public void accept(UniEmitter<? super SecurityIdentity> uniEmitter) {
+                                tenantResolver.getBlockingExecutor().execute(new Runnable() {
+                                    @Override
+                                    public void run() {
+                                        createSecurityIdentityWithOidcServer(uniEmitter, vertxContext, request, resolvedContext,
+                                                userInfo);
+                                    }
+                                });
+                            }
+                        });
+            }
+        } else {
+            return Uni.createFrom().emitter(new Consumer<UniEmitter<? super SecurityIdentity>>() {
+                @Override
+                public void accept(UniEmitter<? super SecurityIdentity> uniEmitter) {
+                    createSecurityIdentityWithOidcServer(uniEmitter, vertxContext, request, resolvedContext, userInfo);
+                }
+            });
+        }
+    }
 
-                                // Token has been verified, as a JWT or an opaque token, possibly involving
-                                // an introspection request.
-                                final TokenCredential tokenCred = request.getToken();
+    @SuppressWarnings("deprecation")
+    private void createSecurityIdentityWithOidcServer(UniEmitter<? super SecurityIdentity> uniEmitter,
+            RoutingContext vertxContext,
+            TokenAuthenticationRequest request, TenantConfigContext resolvedContext, final JsonObject userInfo) {
+        resolvedContext.auth.decodeToken(request.getToken().getToken(),
+                new Handler<AsyncResult<AccessToken>>() {
+                    @Override
```

Comment:
on which thread is this called? Are you 100% sure it cannot be the event loop?

Reply:
where is the blocking call you mentioned earlier?

Reply:
@cescoffier Yeah; so if it is an `IdToken` which is a case with the code flow - then it is `JWT` => hence in 99% cases it is a local verification; if it is a bearer token (the one coming with HTTP `Authorization`) and it is JWT - the same situation. So it is OK to run it on the IO thread.

But if either token is opaque (binary) then it must be a remote introspection call done internally by `Vert.x` as part of this `decodeToken` call - hence in such cases it is a blocking call.

> where is the blocking call you mentioned earlier?

Right now, if OIDC is configured to refresh the token then it will run in the blocking executor starting from [here](https://github.com/quarkusio/quarkus/blob/master/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcIdentityProvider.java#L62), before the decision to verify the token is made, hence this blocking operation control check

Reply:
Vert.x is doing a blocking call and block the event loop? That would be a first. @pmlopes, isn't this hidden?

Reply:
I guess that for you, as a user, you don't see the Vert.x gymnastics.

Reply:
I've no knowledge on the mentioned code, but I can assure that the calls from vert.x auth to the idp server are asynchronous. There's no `executeBlocking` code on vertx-auth-oauth2`

Reply:
HI @cescoffier @pmlopes I think there may be some confusion now, I did not say `Vert.x` does a blocking call, I meant that since `introspection` calls are remote it is better for Quarkus OIDC to run them on the blocking executor, this is why we see so many blocked cases, as in case of the issue this PR is trying to fix, etc

Reply:
@sberyozkin That is a bit confusing. Why "since introspection calls are remote it is better for Quarkus OIDC to run them on the blocking executor"?  Where are these calls? In Vert.x (and so non-blocking) or in some other code?

Reply:
Is it just _easier_ for you to handle the validation? But in this case, how do you make the call (as you may mess up with the event loop used to execute the call and go back to blocking indefinitely).

Reply:
@cescoffier the introspection calls are in Vert.x OAuth2, they are non-blocking (i.e async) but the introspection call is expensive, so we use the Executor for all such remote calls (introspection - starting from this PR, refresh tokens, OIDC connection resolution); IMHO it is OK, perhaps we can revisit it as part of a separate issue, but I believe it is not a bad idea...

Reply:
@cescoffier the previous PR where you helped with splitting the tenant context resolution was related to a remote tenant connection resolution which was run on IO thread in one code branch...even though in Vert.x it is an async call